crates/chia-bls/src/cached_bls.rs

```diff
@@ -1,253 +1,260 @@
+        self.cache.is_empty()
+    }
+
+    pub fn aggregate_verify(
         &mut self,
-        pks: P,
-        msgs: M,
+        pks: impl IntoIterator<Item = impl Borrow<PublicKey>>,
+        msgs: impl IntoIterator<Item = impl AsRef<[u8]>>,
         sig: &Signature,
     ) -> bool {
         let iter = pks.into_iter().zip(msgs).map(|(pk, msg)| -> GTElement {
+            // Hash pubkey + message
             let mut hasher = Sha256::new();
             hasher.update(pk.borrow().to_bytes());
-            hasher.update(msg.as_ref()); // pk + msg
-            let h: [u8; 32] = hasher.finalize().into();
+            hasher.update(msg.as_ref());
+            let hash: [u8; 32] = hasher.finalize().into();
 
-            if let Some(pairing) = self.cache.get(&h).cloned() {
-                // equivalent to `if pairing is not None`
+            // If the pairing is in the cache, we don't need to recalculate it.
+            if let Some(pairing) = self.cache.get(&hash).cloned() {
                 return pairing;
             }
-            // if pairing is None then make pairing and add to cache
+
+            // Otherwise, we need to calculate the pairing and add it to the cache.
             let mut aug_msg = pk.borrow().to_bytes().to_vec();
-            aug_msg.extend_from_slice(msg.as_ref()); // pk + msg
-            let aug_hash: Signature = hash_to_g2(&aug_msg);
+            aug_msg.extend_from_slice(msg.as_ref());
+            let aug_hash = hash_to_g2(&aug_msg);
 
-            let pairing: GTElement = aug_hash.pair(pk.borrow());
             let mut hasher = Sha256::new();
             hasher.update(&aug_msg);
-            let h: [u8; 32] = hasher.finalize().into();
-            self.cache.put(h, pairing.clone());
+            let hash: [u8; 32] = hasher.finalize().into();
+
+            let pairing = aug_hash.pair(pk.borrow());
+            self.cache.put(hash, pairing.clone());
             pairing
         });
-        agg_ver_gt(sig, iter)
+
+        aggregate_verify_gt(sig, iter)
     }
 }
 
-// Python Functions
+#[cfg(feature = "py-bindings")]
+mod python {
+    use super::*;
 
-// Commented out for now as we may remove these
-// as the python consensus code that uses it is being ported to rust.
+    use pyo3::{
+        exceptions::PyValueError,
+        pybacked::PyBackedBytes,
+        pymethods,
+        types::{PyAnyMethods, PyList},
+        Bound, PyResult,
+    };
 
-#[cfg(feature = "py-bindings")]
-#[pymethods]
-impl BLSCache {
-    #[new]
-    pub fn init(size: Option<u32>) -> PyResult<Self> {
-        match size {
-            Some(p_size) => {
-                if p_size < 1 {
-                    Err(PyValueError::new_err(
-                        "Cannot have a cache size less than one.",
-                    ))
-                } else {
-                    Ok(Self::new(NonZeroUsize::new(p_size as usize)))
-                }
-            }
-            None => Ok(Self::default()),
+    #[pymethods]
+    impl BlsCache {
```

Comment:
I imagine this refers the `BlsCache` in the parent module because you `use super::*;` above, right?
Is the reason why you're allowed to add another `impl` block for this type here because you're still in the same file? Because normally a module that doesn't define the type isn't allowed to open an `impl` block for it, as I understand it.

Reply:
You can write an impl for a struct anywhere within the same crate